File: 2020/csharp/Aoc2020/Day02.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace Aoc2020
{
    public class Day02 : IDay
    {
        public override string Day { get; } = nameof(Day02)[3..];

        public override string Result1() => InputLines
            .Select(line => PasswordPolicy.Parse(line))
            .Count(p => p.IsValidOldPassword())
            .ToString();

        public override string Result2() => InputLines
            .Select(line => PasswordPolicy.Parse(line))
            .Count(p => p.IsValidNewPassword())
            .ToString();
    }

    public record PasswordPolicy(string Password, Policy Policy)
    {
        public bool IsValidOldPassword()
        {
            int count = Password.Count(c => c == Policy.Letter);
            return Policy.Min <= count && count <= Policy.Max;
        }

        public bool IsValidNewPassword()
        {
            var (c1, c2) = (Password[Policy.Min - 1], Password[Policy.Max - 1]);
            return (c1 == Policy.Letter || c2 == Policy.Letter) && c1 != c2;
        }

        private static readonly Regex rx = new Regex(@"^(\d+)-(\d+) (.): (.+)$");

        static public PasswordPolicy Parse(string line)
        {
            var matches = rx.Match(line);
            int min = int.Parse(matches.Groups[1].Value);
            int max = int.Parse(matches.Groups[2].Value);
            char letter = matches.Groups[3].Value[0];
            string password = matches.Groups[4].Value;
            Policy policy = new(letter, min, max);
            return new PasswordPolicy(password, policy);
        }
    }

    public record Policy(char Letter, int Min, int Max);
}

File: 2020/fsharp/Aoc2020/Day02.fs
module Day02

open System.Text.RegularExpressions
open Utils

type Policy = { Letter: char; Min: int; Max: int }

type PasswordData = { Password: string; Policy: Policy }

let rx = Regex(@"^(\d+)-(\d+) (.): (.+)$")

let (|PasswordData|_|) input =
    let m = rx.Match(input)
    if (m.Success) then
        let g = m.Groups
        let min = g.[1].Value |> int
        let max = g.[2].Value |> int
        let letter = g.[3].Value.[0]
        let password = g.[4].Value
        Some
            { Password = password
              Policy =
                  { Letter = letter
                    Min = min
                    Max = max } }
    else
        None

let parse line =
    match line with
    | PasswordData pw -> pw
    | _ -> failwith "Invalid input data"

let isValid { Password = password; Policy = { Letter = letter; Min = min; Max = max } } =
    let inRange count = count >= min && count <= max
    password
    |> Seq.filter (fun c -> c = letter)
    |> Seq.length
    |> inRange

let isValid2 { Password = password; Policy = { Letter = letter; Min = min; Max = max } } =
    let c1, c2 = password.[min - 1], password.[max - 1]
    c1 <> c2 && (c1 = letter || c2 = letter)

let job1 () =
    readInputLines "02"
    |> Seq.map parse
    |> Seq.filter isValid
    |> Seq.length
    |> string

let job2 () =
    readInputLines "02"
    |> Seq.map parse
    |> Seq.filter isValid2
    |> Seq.length
    |> string
